subsys/usb/host/usbh_class.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/usb/usbh.h>
+#include <zephyr/logging/log.h>
+
+#include "usbh_class.h"
+#include "usbh_class_api.h"
+#include "usbh_desc.h"
+
+LOG_MODULE_REGISTER(usbh_class, CONFIG_USBH_LOG_LEVEL);
+
+int usbh_class_init_all(struct usbh_context *const uhs_ctx)
+{
+	int ret;
+
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		ret = usbh_class_init(c_node->c_data, uhs_ctx);
+		if (ret != 0) {
+			LOG_ERR("Failed to initialize all registered class instances");
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+int usbh_class_remove_all(const struct usb_device *const udev)
+{
+	int ret;
+
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		if (c_node->c_data->udev == udev) {
+			ret = usbh_class_removed(c_node->c_data);
+			if (ret != 0) {
+				LOG_ERR("Failed to handle device removal for each classes");
+				return ret;
+			}
+
+			/* The class instance is now free to bind to a new device */
+			c_node->c_data->udev = NULL;
+		}
+	}
+
+	return 0;
+}
+
+static int usbh_class_probe_each_iface(struct usbh_class_node *const c_node,
+				       struct usb_device *const udev)
+{
+	const void *const desc_beg = usbh_desc_get_cfg_beg(udev);
+	const void *const desc_end = usbh_desc_get_cfg_end(udev);
+	const struct usb_desc_header *desc = desc_beg;
+	struct usbh_class_data *const c_data = c_node->c_data;
+	struct usbh_class_filter info = {
+		.vid = udev->dev_desc.idVendor,
+		.pid = udev->dev_desc.idProduct,
+	};
+	uint8_t iface;
+	int ret;
+
+	while (true) {
+		desc = usbh_desc_get_next_function(desc, desc_end);
+		if (desc == NULL) {
+			break;
+		}
+
+		if (desc->bDescriptorType == USB_DESC_INTERFACE_ASSOC) {
+			iface = ((struct usb_association_descriptor *)desc)->bFirstInterface;
+		} else if (desc->bDescriptorType == USB_DESC_INTERFACE) {
+			iface = ((struct usb_if_descriptor *)desc)->bInterfaceNumber;
+		} else {
+			__ASSERT(false, "bad implementation of usbh_desc_get_next_function()");
+			return -EINVAL;
+		}
```

Comment:
This part seems duplicated with the follow calling of `usbh_desc_get_iface_info`.